code/game/machinery/doors/airlock.dm

```diff
@@ -737,125 +760,132 @@ About the new airlock wires panel:
-		return STATUS_CLOSE
-	if(!canAIControl() && !isobserver(user))
-		if(canAIHack(user))
-			hack(user)
-		else
-			if(isAllPowerLoss()) //don't really like how this gets checked a second time, but not sure how else to do it.
-				to_chat(user, "<span class='warning'>Unable to interface: Connection timed out.</span>")
-			else
-				to_chat(user, "<span class='warning'>Unable to interface: Connection refused.</span>")
-		return STATUS_CLOSE
+// -------------------------------------------------------------------------------------
 
-	return ..()
 
-/obj/machinery/door/airlock/Topic(href, href_list, nowindow = 0)
-	if(..())
-		return 1
 
-	var/activate = text2num(href_list["activate"])
-	switch(href_list["command"])
-		if("idscan")
-			if(isWireCut(AIRLOCK_WIRE_IDSCAN))
-				to_chat(usr, "The IdScan wire has been cut - IdScan feature permanently disabled.")
-			else if(activate && aiDisabledIdScanner)
-				aiDisabledIdScanner = 0
-				to_chat(usr, "IdScan feature has been enabled.")
-			else if(!activate && !aiDisabledIdScanner)
-				aiDisabledIdScanner = 1
-				to_chat(usr, "IdScan feature has been disabled.")
-		if("main_power")
+/obj/machinery/door/airlock/tgui_act(action, params)
+	if(..())
+		return
+	if(!issilicon(usr) && !usr.can_admin_interact())
+		to_chat(usr, "<span class='warning'>Access denied. Only silicons and admins may use this interface.</span>")
+		return
+	if(issilicon(usr) && emagged)
+		to_chat(usr, "<span class='warning'>Unable to interface: Internal error.</span>")
+		return
+	if(!canAIControl() && !isobserver(usr))
+		if(canAIHack(usr))
+			hack(usr)
+		else
+			if(isAllPowerLoss())
+				to_chat(usr, "<span class='warning'>Unable to interface: Connection timed out.</span>")
+			else
+				to_chat(usr, "<span class='warning'>Unable to interface: Connection refused.</span>")
+		return
+	switch(action)
+		if("disrupt-main")
 			if(!main_power_lost_until)
 				loseMainPower()
 				update_icon()
-		if("backup_power")
+			else
+				to_chat(usr, "<span class='warning'>Main power is already offline.</span>")
+			. = TRUE
+		if("disrupt-backup")
 			if(!backup_power_lost_until)
 				loseBackupPower()
 				update_icon()
-		if("bolts")
-			if(isWireCut(AIRLOCK_WIRE_DOOR_BOLTS))
-				to_chat(usr, "The door bolt control wire has been cut - Door bolts permanently dropped.")
-			else if(activate && lock())
-				to_chat(usr, "The door bolts have been dropped.")
-			else if(!activate && unlock())
-				to_chat(usr, "The door bolts have been raised.")
-		if("electrify_temporary")
-			if(activate && isWireCut(AIRLOCK_WIRE_ELECTRIFY))
+			else
+				to_chat(usr, "<span class='warning'>Backup power is already offline.</span>")
+			. = TRUE
+		if("shock-restore")
+			to_chat(usr, "The door is now un-electrified.")
+			electrify(0)
+			. = TRUE
+		if("shock-temp")
+			if(isWireCut(AIRLOCK_WIRE_ELECTRIFY))
 				to_chat(usr, text("The electrification wire is cut - Door permanently electrified."))
```

Comment:
We dont allow use of `text()`
![image](https://user-images.githubusercontent.com/25063394/88515585-d5d12200-cfe3-11ea-912a-d21783d4b5f4.png)